### PR TITLE
New module: Gitlab webhook

### PR DESCRIPTION
**Pull request by drzraf**
##### SUMMARY
New module: **Gitlab web hooks aka Project hooks**

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
gitlab_webhook

##### ANSIBLE VERSION
```
ansible 2.5.2
```

##### ADDITIONAL INFORMATION
This module provides a way to setup Gitlab webhooks using Ansible.
Contrary to the first 3 gitlab modules (project, user and group), it relies upon the python-gitlab module instead of pyapi-gitlab (see #29310) 

Known bug: The Gitlab webhook "token" parameter is write-only (ie: not exposed on GET request).
That currently keeps module from being idempotent when a task set the `token` parameter.
See https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/7220#note_71782207

### CODE REVIEW

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (gundalow) [inline lib/ansible/modules/source_control/gitlab_webhook.py]: This isn't valid yaml ` : `

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (gundalow) [inline lib/ansible/modules/source_control/gitlab_webhook.py]: I've move all of this into the EXAMPLES block

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (gundalow) [inline lib/ansible/modules/source_control/gitlab_webhook.py]: `required: false` is the default and can be removed

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (andreparames) [inline lib/ansible/modules/source_control/gitlab_webhook.py]: I assume `project.hooks.delete` raises an Exception if it fails. Is the hook ID printed so the user can tell which one failed?

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (andreparames) [inline lib/ansible/modules/source_control/gitlab_webhook.py]: Doesn't the API return any useful information about the error that could be passed to the user?

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (andreparames) [inline lib/ansible/modules/source_control/gitlab_webhook.py]: Bare `except:` is very rarely (if ever) what should be used. In this case, it means a `ctrl-c` will be caught and Ansible will keep going. I suggest using something more specific (at least `Exception`, which won't catch `ctrl-c`).

**Comment by drzraf** [comment]
@andreparames thank you for the review.
I tried to address the issues you mentioned in the last commit. Let me know if you want it to be squashed.

**Review by andreparames — APPROVED** [review]

**Comment by andreparames** [comment]
LGTM. I'll leave the squashing decision to the Ansible core devs.

**Comment by drzraf** [comment]
is there a specific protocol to get merged under 3 days? a ninja-merger somewhere? @maxamillion ?

see also: #40053

**Comment by dj-wasabi** [comment]
@drzraf I don't think there is one. I have closed several of my pr's a while ago after at least half year no response.

shipit

**Comment by maxamillion** [comment]
It would be nice to have some unit tests for this, but I won't block on it.

shipit

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (gundalow) [inline lib/ansible/modules/source_control/gitlab_webhook.py]:
```suggestion
version_added: "2.8"
```

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (gundalow) [inline lib/ansible/modules/source_control/gitlab_webhook.py]:
```suggestion
short_description: Creates, updates or deletes GitLab webhook
```

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (gundalow) [inline lib/ansible/modules/source_control/gitlab_webhook.py]:
```suggestion
  - When the webhook does not exists in this GitLab project, it will be created.
```

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (gundalow) [inline lib/ansible/modules/source_control/gitlab_webhook.py]:
```suggestion
  - When the webhook does exists and C(state=absent), the webhook will be deleted.
```

**Review by gundalow — COMMENTED** [review]
`s/Gitlab/GitLab/g`

**Comment by Shaps** [comment]
Isn't this what the `gitlab_hook` module does?

**Comment by drzraf** [comment]
- This PR was provided May 11, 2018, have gone through the reviewing process and never got considered for merging.
- #40096 was provided May 14, 2018 and merged May 17, 2018

A couple of difference exist like:
- the fact that I tried hard to ensure idempotency, or at least allow for idempotency in case https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/7220#note_71782207 were to be fixed one day.
- the fact that this PR is based on a python library (which is itself tested) to delegate low-level work.

BTW I think the idea of testing against a live GitLab instance (#51490) is the right approach.

**Comment by Lunik** [comment]
Compared to what I have done in #51141, the hook options are better handled in your module.

But now that `gitlab_hooks` is already part of Ansible, we can't remove it and replace it by your module. 

However, we can try to merge your modifications in #51141 ?

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (alxgu) [inline lib/ansible/modules/source_control/gitlab_webhook.py:340]: Why capitalized?
hookHelper or hook_helper

Comment (Lunik) [reply lib/ansible/modules/source_control/gitlab_webhook.py:340]: As long as the input and ouput of the module are in snail_case, I don't see any issue.

**Comment by dagwieers** [comment]
We can still replace or remove the upstream **gitlab_hooks** module. It has not been shipped in an official release yet.

One objection to me is the fact that the module is plural, and we don't allow plural names.

**Comment by dagwieers** [comment]
@Lunik Can we please get your big PR reviewed, tested and merged ASAP ? Because it doesn't help the open PRs if we are refactoring everything at once. IMO we should have first closed all existing PRs before attempting a refactor. Now we have this conflict of interest I'd rather get rid off.

**Comment by Shaps** [comment]
I'd be happy to get the big PR merged but keep this in mind, once the big refactor is complete we can have a closer look at where this module is better and integrate later. We also don't have to necessarily remove the upstream module straightaway, we can alias it and follow the standard deprecation cycle

**Comment by drzraf** [comment]
In terms of methodology I believe two statements are needed:
- whether preference goes to API-backed modules or not
- whether to enforce this preference for new modules

Then a refactor (and new PR reviews) would look like different  according to such a choice.

**Comment by Lunik** [comment]
> In terms of methodology I believe two statements are needed:
> 
> * whether preference goes to API-backed modules or not
> * whether to enforce this preference for new modules
> 
> Then a refactor (and new PR reviews) would look like different according to such a choice.

In the [Community Wiki Guidelines](https://github.com/ansible/community/wiki/Gitlab-Module-Guidelines), we decided to use the `python-gitlab` library for future Gitlab modules. 
So your PR is following working group affinity.

**Comment by drzraf** [comment]
In reply to https://github.com/ansible/community/issues/431#issuecomment-461728960

As commented in the documentation the `solo` parameter exists to comply with least-surprises principle, w.r.t. logic **and** w.r.t. to GitLab API parameters.
Official position of upstream is that webhooks are NOT UNIQUE by their URL because two URL could theoretically be registered with two different tokens. See (and voice) here : https://gitlab.com/gitlab-org/gitlab-ce/issues/39864

About the somehow related idempotency topic, both this module and `gitlab_services` cope with https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/7220 . Best option is to voice/reopen/patch there. As soon as GitLab fixes this, both modules could be made  easily idempotent. I'll add this a ref' to upstream issue at https://github.com/ansible/community/wiki/Gitlab-Module-Guidelines

File: lib/ansible/modules/source_control/gitlab_webhook.py
Comment (Shaps) [inline lib/ansible/modules/source_control/gitlab_webhook.py]: I would rather call this parameter `hook_ssl_verify` or something along those lines since it's related to the hook ssl verification and it doesn't seem very clear to me ( I know it's mentioned in the docs ). Also I don't seem to be able to find where this is used?

**Comment by Shaps** [comment]
This looks good to me. @Lunik, any objections?

shipit

**Comment by Lunik** [comment]
Wait a minute !
What about conflict with `gitlab_hook` module ?

**Comment by Shaps** [comment]
@Lunik I thought since `gitlab_hook` had not officially shipped yet, we could have just merged this and aliased `gitlab_hook` to `gitlab_webhook`?

**Comment by Lunik** [comment]
`gitlab_hook` is in Ansible since 2.6 so I don't understand what you mean ?

**Comment by gundalow** [comment]
https://docs.ansible.com/ansible/devel/modules/gitlab_hook_module.html (singlular, not hook`s`) was added in Ansible 2.6, so should this be closed?

**Comment by drzraf** [comment]
https://github.com/ansible/ansible/pull/39991#issuecomment-459173821

**Comment by gundalow** [comment]
We now have a dedicated directory `/source_control/gitlab/` could the new module please be moved there.

**Comment by Akasurde** [comment]
Hi @drzraf, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>

**Pull request closed without merging**
